src/main/scala/firrtl/analyses/CircuitGraph.scala

```diff
@@ -0,0 +1,137 @@
+// See LICENSE for license details.
+
+package firrtl.analyses
+
+import firrtl.Kind
+import firrtl.annotations.TargetToken.{Instance, OfModule}
+import firrtl.annotations._
+import firrtl.ir.{Circuit, DefInstance}
+
+/** Use to construct [[CircuitGraph]]
+  * Also contains useful related functions
+  */
+object CircuitGraph {
+
+  /** Build a CircuitGraph
+    * [[firrtl.ir.Circuit]] must be of MiddleForm or lower
+    * @param circuit
+    * @return
+    */
+  def apply(circuit: Circuit): CircuitGraph = new CircuitGraph(ConnectionGraph(circuit))
+
+  /** Return a nicely-formatted string of a path of [[firrtl.annotations.ReferenceTarget]]
+    * @param connectionPath
+    * @param tab
+    * @return
+    */
+  def prettyToString(connectionPath: Seq[ReferenceTarget], tab: String = ""): String = {
+    tab + connectionPath.mkString(s"\n$tab")
+  }
+}
+
+/** Graph-representation of a FIRRTL Circuit
+  *
+  * Requires Middle FIRRTL
+  * Useful for writing design-specific custom-transforms that require connectivity information
+  *
+  * @param connectionGraph Source-to-sink connectivity graph
+  */
+class CircuitGraph private[analyses] (val connectionGraph: ConnectionGraph) {
+
+  // Reverse (sink-to-source) connectivity graph
+  lazy val reverseConnectionGraph = connectionGraph.reverseConnectionGraph
+
+  // AST Circuit
+  val circuit = connectionGraph.circuit
+
+  // AST Information
+  val irLookup = connectionGraph.irLookup
+
+  // Module/Instance Hierarchy information
+  lazy val instanceGraph = new InstanceGraph(circuit)
```

Comment:
This is going to race with this PR of mine: https://github.com/freechipsproject/firrtl/pull/1800

So depending on who gets in first, the other will have to fix some things.

Reply:
Sure I'll use `InstanceKeyGraph`.

Reply:
The current version of `InstanceKeyGraph` isn't feature complete. The feature complete version will only be available once #1800 is merged. We could merge #1603 first and then I would be happy to add the change to `InstanceKeyGraph` in #1800.

Reply:
Thank you! If the change takes time, you can just deprecate it and leave the upgrade patch to me.

Reply:
It is already part of https://github.com/freechipsproject/firrtl/pull/1800. So you are good.